Test/git-issues/git-issue-283c.dfy

```diff
@@ -0,0 +1,40 @@
+// RUN: %dafny /compile:0 "%s" > "%t"
+// RUN: %diff "%s.expect" "%t"
+
+datatype Result<T> =
+  | Success(value: T)
+  | Failure(error: string)
+
+datatype Bar = C1() | C2(bl: string)
+
+const X: int := 42;
+const SS: string := "asd"
+
+trait Foo
+{
+  static const S: string := "asd"
+
+  method FooMethod1() returns (r: Result<string>)
+    ensures
+      match Result<string>.Failure(S) {
+        case Failure(X) => true  // X is a constant, but wrong type, so a variable
```

Comment:
In this case X is a variable because there is no declaration of X.
If there is a const declaration, X will not be a variable. The interesting case is if X has an inferred (not declared type). Added implementation to handle this. Simplified the behavior by making X not a variable if there is any static const declaration -- type inference constrains are added and an error reported if there is a type mismatch.